### PR TITLE
Parsers issues in Hoon

### PR DESCRIPTION
**Pull request by mikolajpp**
During the work on the _tune slaw/scot jets_ grant, I have found several 
issues in the Hoon parsers for %da, %if as well a strange behaviour of the hexadecimal number parser. 

## Out-of-range behaviour in %da and %if
%da and %if parsers accept out-of-range inputs, which are nonetheless parsed following the computer calculation logic. It would be better to enforce the range checking directly in the parser and restore %da, %if their proper meaning.

```
> ~2023.5.19..25.300.100..cafe
~2023.5.20..06.01.40..cafe
> .300.50.100.150
.44.50.100.150
```
.

Allowing the value `24` in the hour field of %da is debatable (the meaning would be the first hour of the following day), however I have opted for the hours range `0..23` in the new @da parser.

## Two decrement underflow bugs in %da
In two cases %da parser accepts an input resulting in decrement underflow,
```
~0-.1.1
~292277024402-.1.1
```
.

These result from the structure of the calculation for BC dates; see `+year`. The second of these bugs is triggered for any year past `292.277.024.402` BC.

## Uppercase hexadecimal digits
There is a strange artifact of uppercase hexadecimals at the head of the %ux and %if atoms, but not at their tail,
```
0xcAFE.cafe
.cAFE.a.b.c.0.0.0.0
```
is ok, but 
```
0xcafe.cAFE
``` 
is invalid.

There have actually been a use case of upper-case hexadecimal digits in `sys/zuse.hoon` a while ago, and still is in `urbit-v2.7.pill`, but the `+scap` arm have since been removed and is not present in `develop`. To run this PR on the `2.7` pill, one needs to modify `+scap` arm in `sys/zuse.hoon` to fix the hexadecimal constants.

### CODE REVIEW

**Comment by mikolajpp** [comment]
This PR is coupled with https://github.com/urbit/vere/pull/425. There are new tests in auras.hoon which require new jets to pass.

**Comment by mikolajpp** [comment]
In `app/dbug.hoon` there is an invalid use of unhandled `@x` aura. 
```
...
:~  'cookie'^s+(end [3 4] (rsh [3 2] (scot %x (shax cookie))))
...
```
Unhandled auras are printed in hexadecimal format without the separators that `@ux` has.
In fact, `(scot %z (shax cookie))` would work too, as would any other unknown aura. 
Rather than resort such tricks, we can simply use suitable printer from `co`,
```
:~  'cookie'^s+(end [3 4] (crip ((x-co:co 0) (shax cookie))))
```

I have added a commit to rectify this.

File: pkg/arvo/sys/hoon.hoon [hunk 1]
```diff
@@ -5810,25 +5803,56 @@
...
-          ;~(plug dot dot)
-          ;~  plug
-            dum:ag
-            ;~(pfix dot dum:ag)
-            ;~(pfix dot dum:ag)
-            ;~(pose ;~(pfix ;~(plug dot dot) (most dot qix:ab)) (easy ~))
-          ==
+    |=  tub=nail
+    ^-  (like [[? @] @ @ [@ @ @ (list @ux)]])
+    =<
```
Comment (ashelkovnykov) [inline pkg/arvo/sys/hoon.hoon:5808]: There's a lot of non-standard indentation in this block which can makes it hard to follow at a glance

**Review by ashelkovnykov — CHANGES_REQUESTED** [review]
Mostly LGTM (though I didn't inspect all of the test cases closely because there's so many of them)

File: pkg/arvo/sys/hoon.hoon [hunk 2]
```diff
@@ -5285,7 +5285,7 @@
   ++  piv  (bass 32 (stun [5 5] siv))
   ++  piw  (bass 64 (stun [5 5] siw))
   ++  qeb  (bass 2 ;~(plug seb (stun [0 3] sib)))
-  ++  qex  (bass 16 ;~(plug sex (stun [0 3] hit)))
+  ++  qex  (bass 16 ;~(plug sex (stun [0 3] six)))
```
[the remark this answers is not in the document]
Comment (mikolajpp) [reply pkg/arvo/sys/hoon.hoon:5288]: `+qex` is used in two places: (1) to parse the head of an `@ux` atom and (2) to parse the head of an `@is` atom. 
As described in the issue, I found it strange that uppercase hex are allowed at the head, but not at the tail, and only starting with the second digit. 
```
0xCAFE
```
does not parse, but
```
0xcAFE
```
does. 

This commit proposes that we introduce a uniform notation for the `@ux` and `@is` atoms: they are composed of separated hexadecimal numbers, and only lowercase hex digits are allowed throughout.

File: pkg/arvo/sys/hoon.hoon (hunk 1 above)
Comment (mikolajpp) [reply pkg/arvo/sys/hoon.hoon]: I have improved indentation and comments. Let me know if there are more indentation/style issues.

File: pkg/arvo/sys/hoon.hoon (hunk 2 above)
[the remark this answers is not in the document]
Comment (ashelkovnykov) [reply pkg/arvo/sys/hoon.hoon:5288]: That makes sense. My concern was that it's changing the behaviour of an additional function to fix the behaviour of two downstream functions. However, I no longer think it's a grounded/warranted concern: the alternative would be to change one of the other indirectly called functions which thus solves nothing, and isn't as direct as this change.

File: pkg/arvo/sys/hoon.hoon (hunk 1 above)
Comment (ashelkovnykov) [reply pkg/arvo/sys/hoon.hoon]: Just with regards to spacing & ordering, I'd recommend this:
```
++  when
  ~+
  |=  tub=nail
  ^-  (like [[? @] @ @ [@ @ @ (list @ux)]])
  =>
    |%
    ++  yelq  |=([ad=? yer=@ud] ?:(ad (yelp yer) (yelp (dec yer))))
    ++  two  (bass 10 ;~(plug sed:ab (stun [0 1] sid:ab)))
    ++  duo  (bass 10 ;~(plug sid:ab (stun [0 1] sid:ab)))
    ++  veal
      |=  [a=@ b=?]
      ?:  |(=(0 a) &(=(| b) (gth a +(jes:yo))))
        ~
      (some [b a])
    --
  ::
  =/  vex
  %.  tub
  ;~  plug
      :: year
      ::
      (sear veal ;~(plug dim:ag ;~(pose (cold | hep) (easy &))))
      :: .month
      ::
      ;~(pfix dot (sear |=(m=@ ?:((lte m 12) (some m) ~)) two))
  ==
  ::
  ?~  q.vex
    vex
  =+  yer=->:p.u.q.vex
  =+  mot=+:p.u.q.vex
  ::  day bound
  ::
  =/  mob=@
    ?:  (yelq -:p.u.q.vex)
      (snag (dec mot) moy:yo)
    (snag (dec mot) moh:yo)
  :: .day
  ::
  =/  dex
    %.  q.u.q.vex
    (sear |=(d=@ ?:((lte d mob) (some d) ~)) ;~(pfix dot two))
  ::
  ?~  q.dex
    dex
  :: ..h.m.s..f
  ::
  =/  tex
    %.  q.u.q.dex
    ;~  pose
        ;~  pfix
            ;~(plug dot dot)
            ;~  plug
                (sear |=(h=@ ?:((lte h 23) (some h) ~)) duo)
                ;~(pfix dot (sear |=(m=@ ?:((lte m 59) (some m) ~)) duo))
                ;~(pfix dot (sear |=(s=@ ?:((lte s 59) (some s) ~)) duo))
                ;~(pose ;~(pfix ;~(plug dot dot) (most dot qix:ab)) (easy ~))
        ==  ==
        (easy [0 0 0 ~])
    ==
  ::
  ?~  q.tex
    tex
  :: [year month day time]
  ::
  [p.tex [~ [[-:p.u.q.vex +:p.u.q.vex p.u.q.dex p.u.q.tex] q.u.q.tex]]]
```

You have some really wide expressions that I would usually make taller in *my own* code, but they're still readable and I think the way you have them is on-brand for how the rest of `hoon.hoon` is written.

**Review by ashelkovnykov — CHANGES_REQUESTED** [review]
LGTM pending one style change

**Review by ashelkovnykov — DISMISSED** [review]

**Comment by pkova** [comment]
@ashelkovnykov this looks good to merge, right?

**Comment by joemfb** [comment]
@pkova this has to be coordinated with the linked jet changes to avoid introducing new jet mismatches.

**Comment by ashelkovnykov** [comment]
@pkova Yeah - not sure what you pushed/refactored, but the code looks identical and the approval stands

**Comment by mikolajpp** [comment]
@pkova @ashelkovnykov I have added one more styling commit.

**Review by ashelkovnykov — APPROVED** [review]
